### ISSUE TITLE
Programmatic monitoring of memory usage

### ISSUE THREAD

**Issue by nshepperd** [google/jax#8096]
Jax should provide functions for inspecting the hardware memory usage programmatically. The [device memory profiling](https://jax.readthedocs.io/en/latest/device_memory_profiling.html) feature is sometimes great for debugging, but not useful for certain purposes, such as memory benchmarking of layer implementations, or automatic batch size tuning. IMO, ideally we would have:

- Functions for getting current/peak memory usage per device, like `jax.devices()[0].get_current_bytes_in_use()` or similar.
- A context manager that returns a report including the initial and final bytes in use, as well as peak memory usage during the context, broken out by device. Compare to pytorch's context based cpu/gpu profiling (which is btw another thing it would be really nice to have, again for benchmarking purposes for which tensorboard based profiling is not great).
- On TPUs it would be really nice to know how much of this is padding too...

A more heavyweight version of the above that would be extremely useful would be a context manager that returns a snapshot of metadata about all device arrays resident at peak memory usage, including origin, and information about layout and padding on TPUs, similar to the 'largest arrays' report which is output upon OOM on TPUs (a very useful report that I have been unable to figure out any way of triggering other than deliberately OOMing the device).

See also https://github.com/google/jax/issues/2451

**Comment by josephlr** [comment]
The `memory_stats` method of [`jax.Device`](https://jax.readthedocs.io/en/latest/_autosummary/jax.Device.html#jax-device) should do what you want.

For example, `jax.devices()[0].memory_stats()` returns:
```json
{
  "bytes_in_use": 18874368,
  "bytes_limit": 16624123904,
  "bytes_reservable_limit": 16615718912,
  "bytes_reserved": 16793600,
  "largest_alloc_size": 8388608,
  "largest_free_block_bytes": 16588423168,
  "num_allocs": 11,
  "peak_bytes_in_use": 18907136,
  "peak_bytes_reserved": 16793600
}
```

Looks like this was added in JAX 0.4.12, see https://github.com/google/jax/issues/1491#issuecomment-1588287469

I think this can be closed (unless we want to improve the documentation).

**Comment by vboussange** [comment]
Hey there, `jax.local_devices()[0].memory_stats()` is indeed useful, but how can one reinitialise the `"peak_bytes_in_use"`?

**Comment by mattjj** [comment]
> I think this can be closed

Done!
